kratos/spaces/ublas_space.h

```diff
@@ -277,27 +277,34 @@ class UblasSpace
         return std::sqrt(Dot(rX, rX));
     }
 
-    static TDataType TwoNorm(MatrixType const& rA) // Frobenious norm
+    static TDataType TwoNorm(const Matrix& rA) // Frobenious norm
     {
         TDataType aux_sum = TDataType();
-#ifndef _OPENMP
-        for (int i = 0; i < static_cast<int>(rA.size1()); i++)
-        {
-            for (int j = 0; j < static_cast<int>(rA.size2()); j++)
-            {
-                aux_sum += rA(i,j) * rA(i,j);
+        #pragma omp parallel for reduction(+:aux_sum)
+        for (int i=0; i<static_cast<int>(rA.size1()); ++i) {
+            for (int j=0; j<static_cast<int>(rA.size2()); ++j) {
+                aux_sum += std::pow(rA(i,j),2);
             }
         }
-#else
-        #pragma omp parallel reduction(+:aux_sum)
-        for (int i = 0; i < static_cast<int>(rA.size1()); i++)
-        {
-            for (int j = 0; j < static_cast<int>(rA.size2()); j++)
-            {
-                aux_sum += rA(i,j) * rA(i,j);
+        return std::sqrt(aux_sum);
+    }
+
+    static TDataType TwoNorm(const compressed_matrix<TDataType> & rA) // Frobenious norm
+    {
+        TDataType aux_sum = TDataType();
+
+        const auto& r_row_indices = rA.index1_data();
+        const auto& r_col_indices = rA.index2_data();
+
+        #pragma omp parallel for reduction(+:aux_sum)
```

Comment:
this loop is inefficient because of the access A(i,j)

the efficient way is something like

~~~c++
        #pragma omp parallel for reduction(+:aux_sum)
        for (int i=0; i<static_cast<int>(r_row_indices.size()-1); ++i) {
            const IndexType row_begin = r_row_indices[i];
            const IndexType row_end = r_row_indices[i];

            for (int j=row_begin; j<static_cast<int>(row_end); ++j) {
                aux_sum += std::pow(value_data[j] , 2);
~~~

actually, since you only need to loop over values, and you don't need to know the i,j the most efficient way is

~~~c++
#pragma omp parallel for reduction(+:aux_sum)
for(int i=0; i<value_data.size(); ++i)
    aux_sum += value_data[i];
~~~

Reply:
thx, will apply those changes
I was using `value_data`, not sure why I changed it again :sweat_smile:

Reply:
done